[CBRD-25862] Improve and refactor System Catalog for Information Schema

kangmin5505 opened the pull request:
[<http://jira.cubrid.org/browse/CBRD-25862>
](http://jira.cubrid.org/browse/CBRD-25862) - [구현] System Catalog Class 에 필요한 정보 구현
[<http://jira.cubrid.org/browse/CBRD-26036>
](http://jira.cubrid.org/browse/CBRD-26036)  - [구현] System Catalog Class 에 추가된 내용을 System Catalog View 에 추가


### Purpose
Information Schema 도입 전  CUBRID System Catalog 에 필요한 정보 추가 및 기존 코드를 수정, 제거 및 개선합니다.
- 추가
  - 시간 정보 추가(`created_time`, `updated_time`, `checked_time`): 76231ba571372e3e9c7344a633b948abac26dfbf
  - `serial system catalog class`에 `start_val`  추가: b6b210dc37af0e01188ac7e068db409ebbe89619
  - `user system catalog class`에 `is_loginable`, `is_system_created` 추가: 8c063d7b892328c95871d9fda99362d1a46c9561
  - `stored procedure catalog class` 에 `sql_data_access` 추가: 789378dfae13bd9afce4c5ef5ba37b94fd5c8a8f
  - `partition system catalog class`에 `class_partition_type` 추가: c1b0801fe4b2498bd71e9ee3e0e3b4fe08bdf06f
  - `index system catalog class`에 `referential_match_option`, `referential_index`, `delete_rule`, `update_rule`, `index_type`, `options` 추가: 2a61602a1cf0e47b2a781c259831578f9fb105d2
  - `class system catalog class`에 `statistics_strategy` 추가: 971caa5bffef9cd3d21f30a3308061d04707d5be
  - 추가된 `system catalog class`의 정보를 `system catalog view`에 추가: cfc98916986e7ba7e97f2b4b53d6d8094735bfc1
- 수정
  - `class system catalog class`의 `is_system_class`에서 다른 flag 정보들을 `flags`(컬럼 추가)로 분리: 07bc4a912060fe98330c452c4e5255fa88c44a4f, 9f973f9e18b670b9763549443f20ca6781fdbd3c
- 제거
  - db_authorizations 제거: f4f2d857c4d175066aee76c2bf1727864f565e18
- 개선
  - 성능
    - `system catalog view` 조회 시, 인덱스가 적용되지 않는 부분에 인덱스 추가(`coll_id` - `_db_collation`, `(type_id,  type_name)` - `_db_data_type`, `class_of` - `_db_class`)
  - naming convention
    - system catalog class prefix(\_db\_\*)와 system catalog view prefix(db\_\*)의 naming convention에 맞게 수정: 09128e0e9bb9571eca40bb22e8d8e10eaebdda32

### Remarks
- `return dependency information of PL/CSQL SPs in their compilation results`(dce920d83317cd9a237ed86150920b01f65b7684) 는 `dependency catalog`를 추가하기 위해 필요한 작업입니다. `dependency catalog`도 같이 추가하려고 했지만 자체적으로 테스트할 부분이 남아있어서 시간이 지체되는 관계로, 다른 팀의 catalog 수정 사항을 develop에 반영하기 위한 선행 작업인 이 PR을 우선 반영합니다.
- [b806795](https://github.com/CUBRID/cubrid/pull/6443/commits/b80679522813dd4d014fb2524fca23348e49aa0b) 이후 올라온 commit(fix:)들은 sql, medium 등 테스트를 진행하며 오류를 수정한 것이고, 그 전 commit들은 팀 내 리뷰를 진행했던 commit 입니다.
- **리뷰 완료 시, test branch를 develop 으로 변경**: revert 해야 할 commit(a88c585, 0998a21, 22c3feb)

kangmin5505 committed:
[CBRD-25974] Removing code related to db_authorizations (#6039)

* refactor: remove code related to db_authorizations
* refactor: remove duplicate macro in authenticate_context.cpp and
authenticate_constants.h
* refactor: rename authorizations_class to root_class
* ci: switch test target branch from develop to feature/information_schema

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] Add timestamps to system catalog classes (#6234)

http://jira.cubrid.org/browse/CBRD-25862

- timestamps
  - _db_class: created_time, updated_time, checked_time
  - _db_index: created_time, updated_time
  - _db_stored_procedure: created_time, updated_time
  - _db_server: created_time, updated_time
  - _db_synonym: created_time, updated_time
  - db_user: created_time, updated_time
  - db_serial: created_time, updated_time
  - db_trigger: created_time, updated_time

- refactoring
  - remove unnecessary line
  - correct wrong system catalog name
  - replace hardcoded varchar with format_varchar()
  - replace sm_mark_system_classes with sm_mark_system_class for immediate
marking

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/feature/system-catalog' into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] add timestamp handling code for built-in stored procedure (#6280)

http://jira.cubrid.org/browse/CBRD-25862

- add timestamp handling code from built-in stored procedure

kangmin5505 committed:
[CBRD-25862]  add start_val attribute to db_serial (#6275)

http://jira.cubrid.org/browse/CBRD-25862

start_val: The initial value of the serial when it is first created

kangmin5505 committed:
[CBRD-25862] add is_loginable, is_system_created attributes to db_user (#6276)

http://jira.cubrid.org/browse/CBRD-25862

is_loginable: Indicates whether the user can log in (0 = not allowed, 1 = allowed)
is_system_created: Indicates whether the user was created by the system (0 = user-created, 1 = system-created)

kangmin5505 committed:
[CBRD-25862] add sql_data_access attribute to _db_stored_procedure (#6273)

http://jira.cubrid.org/browse/CBRD-25862

- This attribute describes the type of SQL statements used in the procedure body:
  - NO SQL
  - CONTAINS SQL
  - READS SQL DATA
  - MODIFIES SQL DATA

kangmin5505 committed:
[CBRD-25862] add depth attribute to _db_partition (#6274)

http://jira.cubrid.org/browse/CBRD-25862

- class_partition_type: partition type of class
  - 1: Root partition class
  - 2: Partition class

kangmin5505 committed:
[CBRD-25862] add attributes to _db_index (#6287)

http://jira.cubrid.org/browse/CBRD-25862

- Added attributes to _db_index system class:
  - referential_match_option (INTEGER): foreign key match type (NONE, PARTIAL, FULL)
  - referential_index (_DB_INDEX): referenced index for foreign key
  - delete_rule (INTEGER): ON DELETE action (CASCADE, SET NULL, RESTRICT, NO ACTION)
  - update_rule (INTEGER): ON UPDATE action (CASCADE, SET NULL, RESTRICT, NO ACTION)
  - index_type (INTEGER): type of index (currently only BTREE supported)
  - options (INTEGER): index options (currently deduplicate_key_level only)
- Added helper function: get_class_constraint_att_count() to get number of columns in a foreign key
- Added helper function: get_class_constraint_index() to calculate field index in foreign key info sequence using enum and sequence size
- Replaced magic numbers and cleaned up related code

kangmin5505 committed:
[CBRD-26036] Reflect system class attributes in system catalog views (#6331)

http://jira.cubrid.org/browse/CBRD-26036

- db_class(created_time, updated_time, checked_time)
- db_index(referential_index_class_owner_name, referential_index_class_name, referential_index_name, delete_rule, update_rule, referential_match_option, index_type, deduplicate_key_level created_time, updated_time) - db_partition(class_partition_type)
- db_stored_procedure(created_time, updated_time)
- db_server(created_time, updated_time)
- db_synonym(created_time, updated_time)
- db_trig(created_time, updated_time)

kangmin5505 committed:
[CBRD-25862] Add statistics_strategy column to class system catalog (#6341)

http://jira.cubrid.org/browse/CBRD-25862

Refactoring
- Add `CT_ATTR_CLASS_INDEX` enum and replace literal strings with enum in `CT_CLASS`
Fix
- Prevent statistics updates for views
- Update system catalog class list in `sm_update_all_catalog_statistics to match the CUBRID manual:
  - Add `db_dual`
  - Remove `db_password`

kangmin5505 committed:
[CBRD-25862] Improve performance of system catalog view query (#6318)

http://jira.cubrid.org/browse/CBRD-25862

- Add an index on `coll_id` in `_db_collation` for `db_class`
- Add an index on (`type_id`, `type_name`) in `_db_data_type` for `db_attribute`
- Add an index on `class_of` in `_db_class` for `db_trigger`

kangmin5505 committed:
[CBRD-25862] System catalog naming convention (#6307)

http://jira.cubrid.org/browse/CBRD-25862

Rename system catalog classes:
- `db_ha_apply_info` -> `_db_ha_apply_info`
- `db_serial` -> `_db_serial`
- `db_trigger` -> `_db_trigger`

Rename and created system catalog views:
- Created view `db_ha_apply_info`
- Created view `db_serial`
- Renamed view `db_trig` -> `db_trigger`

Remarks:
- Authorization classes (`db_authorization`, `db_password`, `db_root`, `db_user`) are not changed because they are handled separately from system catalog classes. They may be reviewed for changes in the future if necessary.

kangmin5505 committed:
[CBRD-25941] rename is_system_class to flags (#6376)

http://jira.cubrid.org/browse/CBRD-25941

Rename is_system_class to flags

hyunikn committed:
Merge branch 'develop' into feature/system-catalog

kangmin5505 committed:
[CBRD-25765] Refactor system catalog function namings (#6393)

http://jira.cubrid.org/browse/CBRD-25765

- Rename CT_DB_* and CTV_DB_* constants to CT_* and CTV_* (drop redundant DB_ prefix)
- Standardize catalog initializer methods and view spec names (e.g. get_method_sig → get_meth_sig, get_class_authorization → get_auth, get_view_super_class → get_view_direct_super_class)
- Clean up includes and unused variables

hyunikn committed:
[CBRD-26248] return dependency information of PL/CSQL SPs in their compilation results (#6390)

* add depencies data to the responses to PL/CSQL compile requests

* add dependencies information into the response message to the Static SQL semantics check request,
and include the dependencies into the response of PL/CSQL compile request

kangmin5505 committed:
Merge branch 'develop' into feature/system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] System catalog naming convention - revert contrib/php4/src/php_cubrid.c (#6419)

revert: contrib/php4/src/php_cubrid.c

kangmin5505 committed:
fix: use correct DB_VALUEs on classobj_put_foreign_key_ref()

kangmin5505 committed:
fix: check authorization when accessing to index catalog class

kangmin5505 committed:
fix: pass to check authorization when accessing _db_serial

kangmin5505 committed:
[CBRD-26248] fix when no dependencies on PR(return dependency information of PL/CSQL SPs in their compilation results) (#6428)

http://jira.cubrid.org/browse/CBRD-26248

- addAll when dependencies exist in static sql

kangmin5505 committed:
Merge branch 'feature/system-catalog' into fix_system-catalog

kangmin5505 committed:
fix: replace is_system_class with flags about _db_class

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
Merge branch 'fix_system-catalog' into feature/system-catalog

kangmin5505 committed:
codestyle

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
fix: pass to check authorization when accessing _db_serial

kangmin5505 committed:
temp: change test branch from develop to feature/system-catalog

kangmin5505 committed:
codestyle

kangmin5505 commented:
/run all

kangmin5505 committed:
temp: change test branch from develop to feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 commented:
/run all

kangmin5505 committed:
fix: disable authorization check when querying serial

kangmin5505 commented:
/run all

kangmin5505 committed:
fix: disable authorization check when querying serial

kangmin5505 commented:
/run shell

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
fix: disable authorization check when accessing serial

kangmin5505 commented:
/run shell

kangmin5505 commented:
/run shell

kangmin5505 commented:
/run shell

kangmin5505 commented:
/run all

tw-kang commented:
/run all

kangmin5505 commented:
/run all

kangmin5505 commented:
/run all

kangmin5505 commented:
/run all

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 committed:
fix: change CATCLS_USER_ATTR_IDX_NAME from 7 to 11

kangmin5505 commented:
/run all

kangmin5505 committed:
fix: replace _db_* with db_* for trigger, serial

kangmin5505 commented:
/run all

kangmin5505 commented:
/run shell

kangmin5505 commented:
/run shell

kangmin5505 committed:
[CBRD-25862] Move trigger system catalog install logic from trigger manager to schema system catalog module (#6451)

http://jira.cubrid.org/browse/CBRD-25862

move trigger system catalog to schema_system_catalog_install

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 commented:
/run all

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
현재 여러 세션에서 통계 정보를 갱신하는 경우 deadlock 이 발생하는 문제가 있습니다.
그 부분은 제외하고 확인해 주시면 감사하겠습니다.

kangmin5505 commented:
/run all

kangmin5505 commented on src/object/schema_template.c:
```diff
@@ -4962,18 +4962,29 @@ smt_change_constraint_status (SM_TEMPLATE * ctemplate, const char *index_name, S
+  MOP index_catalog_class = NULL;
+  int save;
+
+  AU_DISABLE (save);
```
_db_index(system table) 의 값을 가져오기 위한 권한 권한 검사 해제
(신규 함수)

kangmin5505 commented on src/query/execute_statement.c:
```diff
@@ -3017,6 +3017,9 @@ do_drop_serial (PARSER_CONTEXT * parser, PT_NODE * statement)
       goto end;
     }
 
+  AU_DISABLE (save);
+  au_disable_flag = true;
```
_db_serial(system table ) 에서 값을 가져오기 위한 권한 검사 해제
(기존: 일반 유저도 serial system table 대한 SELECT 권한이 있었음)

kangmin5505 commented on src/executables/unload_schema.c:
```diff
@@ -741,14 +741,14 @@ export_serial (extract_context & ctxt, print_output & output_ctx)
   db_make_null (&diff_value);
   db_make_null (&answer_value);
 
+  AU_DISABLE (save);
```
_db_serial(system table)에 쿼리를 하기 위한 권한 검사 해제
(기존: 일반 유저도 serial system table 대한 SELECT 권한이 있었음)

shparkcubrid commented:
```
class system catalog class의 is_system_class를 flags로 변경
```
컬럼명을 수정할 필요가 있을까요? 카탈로그 테이블은 내부에서만 사용하지 않고 외부에서도 사용될 수 있습니다. 하위호환성을 깰 필요가 있는지 궁금합니다.

kangmin5505 commented:
> ```
> class system catalog class의 is_system_class를 flags로 변경
> ```
> 
> 컬럼명을 수정할 필요가 있을까요? 카탈로그 테이블은 내부에서만 사용하지 않고 외부에서도 사용될 수 있습니다. 하위호환성을 깰 필요가 있는지 궁금합니다.
컬럼명을 수정해야겠다고 판단한 이유는 아래와 같습니다.
1. is_system_class 변수에 is_system_class 이외의 플래그 정보 포함(ex. reuse_oid)
2. 카탈로그 테이블은 시스템 내부에서 사용되는 정보로 외부에서는 카탈로그 뷰를 사용하는 것이 적절하다고 판단

1번에 대해서는 is_system_class 를 유지하고 다른 변수를 생성하자는 의견도 있었습니다. 하지만 이미 is_system_class 변수를 활용하여 reuse oid  인지를 확인하는 코드(ex. is_system_class & 8 => reuse oid 관련 정보 추출)도 있었고, 불필요한 저장공간을 차지(flag 성 정보에 4byte 할당)한다고 생각하였습니다.

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
@shparkcubrid 
`_db_class`의 `is_system_class` 컬럼은 유지하고 `is_system_class`에 있는 다른 flag 값들은 `flags` 컬럼을 추가하여 관리하기로 하였습니다. 해당 PR 시, reviewer 등록하겠습니다.

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] Fix deadlock of updating statistics (#6546)

http://jira.cubrid.org/browse/CBRD-25862

- fix: deadlock of updating statistics
- refactor
  - unify duplicated logic for statistics update
  - replace finding classes query with internal API

kangmin5505 committed:
Merge remote-tracking branch 'upstream/develop' into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] Revert renaming is_system_class in _db_class and add new column(flags) (#6551)

http://jira.cubrid.org/browse/CBRD-25862

- rename gathering strategy index to statistics strategy index
- split legacy packed flags of _db_class for compatibility

kangmin5505 committed:
Merge branch 'feature/system-catalog' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] Refactor to set null initial value of checked time (#6580)

http://jira.cubrid.org/browse/CBRD-25862

- refactor: simplify db_string_to_datetime based on db_string_to_datetime2
- feature: print NULL if not updated statistics

kangmin5505 committed:
temp: change test branch from develop to feature/system-catalog

kangmin5505 committed:
Merge branch 'feature/system-catalog' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 commented:
/run sql medium

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 committed:
[CBRD-25862] Remove timestamp information and statistics_strategy on META_CLASS (#6632)

http://jira.cubrid.org/browse/CBRD-25862

* refactor: remove duplicated PT_NAME_ORIGINAL macro
* refactor: move _db_class timestamp update logic from CAS to cub_server
* refactor: move _db_class checked_time and statistics_strategy update logic from CAS to cub_server
* refactor: move external variables and functions to header file
* refactor: move _db_index timestamp update logic from CAS to cub_server

* fix: acquire IS_LOCK of class before updating statistics

* performance: add to cache fixed attribute

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 commented:
sql 테스트 실패(2건)는 해당 PR과 무관한 케이스(다른 PR에서도 발생 중 - https://github.com/CUBRID/cubrid/pull/6484)이고, 
shell 테스트 실패(3건)는 flaky 입니다.
manual test 통과 이후 develop에 머지하도록 하겠습니다.

hyunikn commented:
> sql 테스트 실패(2건)는 해당 PR과 무관한 케이스(다른 PR에서도 발생 중 - [https://github.com/CUBRID/cubrid/pull/6484)이고](https://github.com/CUBRID/cubrid/pull/6484)%EC%9D%B4%EA%B3%A0), shell 테스트 실패(3건)는 flaky 입니다. manual test 통과 이후 develop에 머지하도록 하겠습니다.

develop 의 마지막 commit 8c471d1a8d4 에서는 모든 테스트 fail 없이 성공했습니다. (https://app.circleci.com/pipelines/github/CUBRID/cubrid?branch=pull%2F6571%2Fhead)
여기까지 merge 해서 다시 돌리면 fail 갯수가 줄어들지도 모릅니다.

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

hyunikn commented:
실패가 Shell Test 한 건으로 줄었네요. 
로컬에서 한번 실행해 보고 재현 안되면 unstable 이니 명환님께 강제 머지를 부탁드릴 수 있겠습니다. 
flaky 라고 마킹되어 있으니 로컬 재현도 필요 없으려나요.

kangmin5505 commented:
@hyunikn 
확인 감사합니다.
해당 flaky 건과 관련하여 PR이 진행 중인 거 같습니다. 해당 PR이 머지되면 다시 수행해 보도록 하겠습니다.
https://github.com/CUBRID/cubrid-testcases-private-ex/pull/2891

kangmin5505 commented:
/run shell

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 committed:
fix: update constraint_seq size requirement in or_install_btids_constraint function

kangmin5505 committed:
Merge branch 'develop' of github.com:CUBRID/cubrid into feature/system-catalog

kangmin5505 commented:
/run all

kangmin5505 committed:
Revert "temp: change test branch from develop to feature/system-catalog"

This reverts commit 22c3feb3c4a6f1835ba3e960c4281369ea809faa.

kangmin5505 committed:
Revert "temp: change test branch from develop to feature/system-catalog"

This reverts commit 0998a21e49b7a7fd3ac764c9b5d0f9b2066962e7.

kangmin5505 committed:
Revert "temp: change test branch from develop to feature/system-catalog"

This reverts commit a88c58522602bfc26ec197ccd280bb07748dbd4a.

kangmin5505 committed:
Merge branch 'develop' into feature/system-catalog
